hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+
+Because the hash function can quickly convert a key to an array index, this operation can result in very quick lookup times.
+
+<!-- Lucidchart link  https://www.lucidchart.com/invitations/accept/5fdcf503-7d8b-4139-94d4-795bfed27883 -->
+
+![Hash Function Example](../assets/hash-tables_hash-tables_hash-function.png)
+
+<!-- Hash function image source: https://en.wikipedia.org/wiki/Hash_function -->
+
+### !callout-info
+
+## Hash Tables Implement Associative Arrays
+
+One reason why hash tables are interesting is because they implement an _abstract data type_ known as an _associative array_. An _abstract data type_ is any data structure that is theoretical, and described not by its concrete implementation, but by its behavior and how it's used. An _associative array_ is a theoretical data structure that describes collections that store key-value pairs.
+
+### !end-callout
+
+## Time Complexity
+
+For all three operations of inserting, finding, and deleting values, hash tables have an O(1) time complexity.
+
+| Action                        | Average Case | Worst Case |
+| ----------------------------- | ------------ | ---------- |
+| Finding a value by key        | O(1)         | O(n)       |
+| Removing a value by key       | O(1)         | O(n)       |
+| Inserting a new key and value | O(1)         | O(n)       |
+
+### !callout-info
+
+## We Describe Hash Table Big O by Average Case
+
+Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average-case**. The likelihood of the worst case from occurring depends on how well the **hash function** is implemented.
+
+### !end-callout
+
+## Hash Function
+
+**Hash functions** are functions responsible for mapping a given key into an integer index in the internal array.
+
+### Example Hash Function
+
+Imagine that we have a hash table which holds "name"-"job title" key-value pairs. This hash table needs a hash function to transform the "name" into an index.
+
+An example hash function could use the following algorithm:
+
+- Convert the key to a string
+- Grab the length of this string
+- Get modulo of the length of the string and the length of the internal array
+- Use this modulo as the index
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+### !callout-info
+
+## Why Modulo in This Example?
+
+When arrays get created, they get created with a fixed size in memory. This means that they have a maximum length and _can't_ hold an infinite number of things (unless there is an operation which expands the array. Python's list method `append` manages this for us!). Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!
```

Comment:
I think it will keep things simpler to avoid merging this discussion with thinking about how arrays "grow". Let's pretend that doesn't exist. 🙃
removed a little redundancy
add an additional tidbit about how often, there is a part of the calculation handled by the key type itself, and a part handled by the table
```suggestion
When arrays get created, they have a fixed size in memory. Trying to read or write beyond the end would result in an `IndexError`. Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!

<br />

In practice, the class representing the key type will be responsible for converting the key to some initial numeric value (Python classes can implement the `__hash__` method for this purpose), and the hash table performs the additional steps to make sure the numeric value refers to a valid index in its internal storage.
```